synapse/storage/databases/main/event_push_actions.py

```diff
@@ -1144,6 +1265,64 @@ def _handle_new_receipts_for_notifs_txn(self, txn: LoggingTransaction) -> bool:
                 value_values=[(row[0], row[1]) for row in unread_counts],
             )
 
+        # For each new read receipt we delete push actions from before it and
+        # recalculate the summary.
+        for _, room_id, user_id, thread_id, stream_ordering in rows:
```

Comment:
The contents of this loop are *similar* to the above, but also a bit different. It might be worth trying to combine them, but could get messier?